<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Откройте-таблицу-и-изучите-общую-информацию-о-данных" data-toc-modified-id="Откройте-таблицу-и-изучите-общую-информацию-о-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Откройте таблицу и изучите общую информацию о данных</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Удаление-пропусков" data-toc-modified-id="Удаление-пропусков-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Удаление пропусков</a></span></li><li><span><a href="#Обработка-аномальных-значений" data-toc-modified-id="Обработка-аномальных-значений-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Обработка аномальных значений</a></span></li><li><span><a href="#Удаление-пропусков-(продолжение)" data-toc-modified-id="Удаление-пропусков-(продолжение)-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Удаление пропусков (продолжение)</a></span></li><li><span><a href="#Изменение-типов-данных" data-toc-modified-id="Изменение-типов-данных-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Изменение типов данных</a></span></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Обработка дубликатов</a></span></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Категоризация данных</a></span></li><li><span><a href="#Шаг-3.-Исследуйте-данные-и-ответьте-на-вопросы" data-toc-modified-id="Шаг-3.-Исследуйте-данные-и-ответьте-на-вопросы-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>Шаг 3. Исследуйте данные и ответьте на вопросы</a></span><ul class="toc-item"><li><span><a href="#3.1-Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="3.1-Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?-2.7.1"><span class="toc-item-num">2.7.1&nbsp;&nbsp;</span>3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?</a></span></li><li><span><a href="#3.2-Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="3.2-Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-2.7.2"><span class="toc-item-num">2.7.2&nbsp;&nbsp;</span>3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#3.3-Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="3.3-Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-2.7.3"><span class="toc-item-num">2.7.3&nbsp;&nbsp;</span>3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#3.4-Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="3.4-Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-2.7.4"><span class="toc-item-num">2.7.4&nbsp;&nbsp;</span>3.4 Как разные цели кредита влияют на его возврат в срок?</a></span></li><li><span><a href="#3.5-Приведите-возможные-причины-появления-пропусков-в-исходных-данных." data-toc-modified-id="3.5-Приведите-возможные-причины-появления-пропусков-в-исходных-данных.-2.7.5"><span class="toc-item-num">2.7.5&nbsp;&nbsp;</span>3.5 Приведите возможные причины появления пропусков в исходных данных.</a></span></li><li><span><a href="#3.6-Объясните,-почему-заполнить-пропуски-медианным-значением-—-лучшее-решение-для-количественных-переменных." data-toc-modified-id="3.6-Объясните,-почему-заполнить-пропуски-медианным-значением-—-лучшее-решение-для-количественных-переменных.-2.7.6"><span class="toc-item-num">2.7.6&nbsp;&nbsp;</span>3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.</a></span></li></ul></li><li><span><a href="#Шаг-4:-общий-вывод." data-toc-modified-id="Шаг-4:-общий-вывод.-2.8"><span class="toc-item-num">2.8&nbsp;&nbsp;</span>Шаг 4: общий вывод.</a></span></li></ul></li></ul></div>

# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



**Цель исследования** — определить какие параметры влияют на факт погашения кредита в срок.

Необходимо проверить следующие зависимости:
1. Есть ли зависимость между количеством детей и возвратом кредита в срок?
2. Есть ли зависимость между семейным положением и возвратом кредита в срок?
3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
4. Как разные цели кредита влияют на его возврат в срок?

**Описание проекта**

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

<p>
<div class="alert-success">
<b>Комментарий ревьюера ✔️</b>

Здорово и необычно, что ты дополняешь первую часть проекта. Вступление станет лучше, если ты добавишь в него описание данных, тогда получатель отчета сможет быстрей его понять. 



Пример описания данных:
- *airports_nearest* — расстояние до ближайшего аэропорта в метрах (м)
- *balcony* — число балконов
- ...

    
Твой проект станет намного удобней в использовании, если ты добавишь интерактивное оглавление — с помощью гиперссылок или в плагине toc. Материал об этом есть в блоке курса Полезные инструкции для учёбы - Оформление проекта - Изменения в Jupyter: автоматическое оглавление и горячие клавиши.

## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

<p>
<div class="alert-success">
<b>Комментарий ревьюера ✔️</b>


Блок загрузки библиотек лучше размещать в отдельной от остального кода ячейке. Такой подход позволит улучшить читаемость кода, снизит время поиска ошибок (если они есть) и исключит повторное выполнение кода.
    
Здорово, что здесь используется конструкция ``try-except`` для путей файлов. Но еще лучше использовать библиотеку ``os``:

    import os

    pth1 = '/folder_1/data.csv'
    pth2 = '/folder_2/data.csv'
    
    if os.path.exists(pth1):
        query_1 = pd.read_csv(pth1)
    elif os.path.exists(pth2):
        query_1 = pd.read_csv(pth2)
    else:
        print('Something is wrong)

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


В таблице 12 столбцов. Тип данных в двух столбцах - float; в пяти столбцах - int; в пяти столбцах — object.

Согласно документации к данным:
Описание данных
* `children` — количество детей в семье
* `days_employed` — общий трудовой стаж в днях
* `dob_years` — возраст клиента в годах
* `education` — уровень образования клиента
* `education_id` — идентификатор уровня образования
* `family_status` — семейное положение
* `family_status_id` — идентификатор семейного положения
* `gender` — пол клиента
* `income_type` — тип занятости
* `debt` — имел ли задолженность по возврату кредитов
* `total_income` — ежемесячный доход
* `purpose` — цель получения кредита

## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

В столбце `'days_employed'` и `'total_income'` пропущено 2174 значений.

<p>
<div class="alert-success">
<b>Комментарий ревьюера ✔️</b>


Удобно было бы сразу вывести долю пропусков. Так мы увидим, что можно отбросить, а с чем надо работать. Пример кода:
    
    pd.DataFrame(data.isna().mean()*100).style.set_precision(1).background_gradient('coolwarm')

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
# Создадим цикл, в котором пройдем по каждому уникальному типу занятости (столбец 'income_type')
# Далее в столбце 'income_type' в строки с пропуском запишем медианное значение 
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

<p>
<div class="alert-success">
<b>Комментарий ревьюера ✔️</b>

Сейчас доля аномалий мала и ни на что не повлияет. В следующий раз старайся сразу высчитывать их долю. Мы можем удалить не больше 5-10% датафрейма. Если доля удаляемых данных больше, надо обрабатывать их по-иному

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

<p>
<div class="alert-success">
<b>Комментарий ревьюера ✔️</b>



Можно было бы еще обработать странное значение в данных пола. Оно единично и ни на что не повлияет, но описать его следует

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
# Создадим цикл, в котором пройдем по каждому уникальному типу занятости (столбец 'income_type')
# Далее в столбце 'income_type' в строки с пропуском запишем медианное значение общего трудового стажа ('days_employed')

for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

В каждой подстроке для каждой цели можно выделить определенные повторяющиеся слова. 

Например, для категории 'операции с автомобилем' будет подстрока 'автом'. Для категории 'операции с недвижимостью' - подстрока 'жил'. Для категории 'проведение свадьбы' - подстрока 'свад'. Для категории 'получение образования' - подстрока 'образов'.

По этим подстрокам в функции будет производиться присваивание категории.

In [20]:
def categorize_purpose(row):
    '''
    Функция ищет в столбце ключевы слова ('автом', 'жил', 'свад', 'образов') и по ним возвращает название категории для цели кредита
    '''
    
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

<p>
<div class="alert-success">
<b>Комментарий ревьюера ✔️</b>
    
Каждый шаг работы стоит заканчивать кратким выводом.

### Шаг 3. Исследуйте данные и ответьте на вопросы

<p>
<div class="alert-info">
<b>Комментарий студента</b>
    
Так как в первом варианте кода я посчитала не то, что нужно, то я решила удалить данный код, чтобы он не мешал.
    
Ниже привожу решение согласно замечаниям.

    

</div>

<p>
<div class="alert-success">
<b>Комментарий ревьюера 2 ✔️</b>

    
Отлично 👍 </div>

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

<p>
<div class="alert-info">
<b>Комментарий студента</b>
    
Для начало посмотрим, распределены ли равномерно выборки, если сгруппировать семьи по количеству детей.
    
Для этого сгруппирую семьи по количеству детей (столбец 'children') и посчитаю количество семей, которые брали кредит. Для этого применю агрегирующую функцию count к столбцу 'debt'.
    
</div>

In [22]:
data.groupby('children')['debt'].agg('count')

children
0    14091
1     4808
2     2052
3      330
4       41
5        9
Name: debt, dtype: int64

<p>
<div class="alert-info">
<b>Комментарий студента</b>
    
Больше всего данных представлено по семьям с 0, 1 и 2 детьми. Чтобы выявить зависимость между количеством детей и возвратом кредита в срок будем рассматривать именно эти семьи. Семьи с 3 и более детьми не будем включать в анализ, так как по ним имеется мало данных.
    
</div>

<p>
<div class="alert-info">
<b>Комментарий студента</b>
    
Создадим сводную таблицу, где:
* строки - количество детей в семье; 
* колонка 'count debt' - число кредитополучателей;
* колонка 'sum debt' - количество должников;
* колонка 'mean debt' - доля должников.
    
</div>

In [23]:
data_pivot_table = data.pivot_table(index='children', values='debt', aggfunc=['count', 'sum', 'mean'])

# Выведем информацию только по семьям с 0, 1 и 2 детьми
data_pivot_table.head(3)

count   sum      mean
           debt  debt      debt
children                       
0         14091  1063  0.075438
1          4808   444  0.092346
2          2052   194  0.094542

<p>
<div class="alert-info">
<b>Комментарий студента</b>
  
**Вывод:**

Доля должников в семьях без детей составляет всего 7,54%, а в семьях с 1, 2 детьми уже 9%. Возможно семьи без детей чаще выплачивают кредит в срок, так как у нет ежедневных трат на детей.

Можно сделать вывод о том, что чем меньше детей в семье, тем выше кредитоспособность семьи.
    
      
</div>

<p>
<div class="alert-danger">
<b>Комментарий ревьюера ❌</b> 
    
Проделана большая работа! Однако такое большое количество разных табличек сложно воспринимается, да и посчитано не совсем то, что нужно. В этом проекте важно потренироваться делать сводные таблицы `pivot_table()`. Выявить зависимость можно сравнив только самые большие выборки с 0, 1 и 2 детьми.
    
Нужно сделать `pivot_table()`, где строки (index) – категория, в данном случае количество детей, значения (values) в нашем случае – это debt, а функции (aggfunc) нужно сделать три. Первая будет `count` – число строк, она же число всех кредитополучателей, вторая `sum` – просуммирует единицы, то есть посчитает количество должников, а третья – `mean` – доля должников. Из-за того, что значения только нули и единицы, среднее будет отношением количества единиц (должников) к количеству всех строк. Для того, чтобы сделать несколько агрегирующих функций, нужно перечислить их в квадратных скобках aggfunc=[...]

<p>
<div class="alert-success">
<b>Комментарий ревьюера 2 ✔️</b>


    
- Верно 👍 Кредитополучатели без детей являются немного более ответственными, чем клиенты у которых есть дети. 
  
    
- Для того, чтобы таблица смотрелась лучше стоит переименовать столбцы (здесь и далее), вот пример кода, как это можно сделать:
  
    
    data.columns = ['Всего кредитополучателей', 'Всего должников', 'Доля должников']
    
- Можно сделать универсальную функцию для этого и следущих пунктов. 
    

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

<p>
<div class="alert-info">
<b>Комментарий студента</b>
    
Создадим сводную таблицу, где:
* строки - семейное положение; 
* колонка 'count debt' - число кредитополучателей;
* колонка 'sum debt' - количество должников;
* колонка 'mean debt' - доля должников.
    
</div>

In [24]:
data.pivot_table(index='family_status', values='debt', aggfunc=['count', 'sum', 'mean'])

count  sum      mean
                        debt debt      debt
family_status                              
Не женат / не замужем   2796  273  0.097639
в разводе               1189   84  0.070648
вдовец / вдова           951   63  0.066246
гражданский брак        4134  385  0.093130
женат / замужем        12261  927  0.075606

<p>
<div class="alert-info">
<b>Комментарий студента</b>
 
Наибольшая доля должников - люди не состоящих в браке (9.8%) и семьи в гражданском браке (9.3%).
  
Меньше всего должников среди вдовцов/вдов (6.6%), людей в разводе (7.1%) и состоящих в браке (7.6%). 
    
Однако стоит отметить перекос в группах. Больше всего кредит берут семьи состоящие в браке, их почти 12 300. На втором месте люди не состоящих в браке (их почти 2800) и люди в разводе (почти 1200). Несмотря на то, что меньше всего должников среди вдовцов/вдов, эти люди реже всех берут кредиты.


<div class="alert-warning">
<b>Комментарий ревьюера ⚠️</b>

Комментарии лучше писать в ячейках Markdown, а через # пишут небольшие технические комментарии к коду. 

<p>
<div class="alert-danger">
<b>Комментарий ревьюера ❌</b> 
    
Нужно сделать сводную таблицу, где строки – категория, значения – это наличие долга, а столбцы – число всех кредитополучателей, количество должников и доля должников. Именно по доле должников мы и делаем дальнейшие выводы. При этом нужно так же оценить размеры групп, нет ли перекоса.

<p>
<div class="alert-info">
<b>Комментарий студента</b>
    
**Вывод:** 

Меньше всего должников среди людей в разводе и состоящих в браке. Также редкие должники вдовцы/вдовы, однако они и не часто берут кредиты. Чаще всех кредиты берут люди в браке.
 

</div>

<p>
<div class="alert-success">
<b>Комментарий ревьюера 2 ✔️</b>

    
Согласна, по расчетам можно вывести общую зависимость, что заемщики состоящие или состоявшие в браке лучше выплачивают кредиты.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

В датафрейме был создан ноывй столбец `total_income_category`, в котором для удобства величина дохода была разбита на категории.

В столбце `total_income_category` выделены следующие категории в зависимости от дохода кредитополучателя:

* `'E'` доход 0 - 30 000;
* `'D'` доход 30 001 - 50 000;
* `'C'` доход 50 001 - 200 000;
* `'B'` доход 200 001 - 1 000 000;
* `'A'` доход выше 1 000 000.

<p>
<div class="alert-info">
<b>Комментарий студента</b>
    
Создадим сводную таблицу, где:
* строки - категория в зависимости от дохода кредитополучателя; 
* колонка 'count debt' - число кредитополучателей;
* колонка 'sum debt' - количество должников;
* колонка 'mean debt' - доля должников.
    
</div>

In [25]:
data.pivot_table(index='total_income_category', values='debt', aggfunc=['count', 'sum', 'mean'])

count   sum      mean
                        debt  debt      debt
total_income_category                       
A                         25     2  0.080000
B                       5014   354  0.070602
C                      15921  1353  0.084982
D                        349    21  0.060172
E                         22     2  0.090909

<p>
<div class="alert-info">
<b>Комментарий студента</b>
    
Выборки крайне несбалансированны и для выявления зависимости между уровнем дохода и возвратом кредита в срок стоит сравнить две самые большие категории B и C.
    
Доля должников среди людей категории B с доходом 200 001 - 1 000 000 составляет 7%, а доля должников с меньшим доходом категория C (50 001 - 200 000) уже 8.5%.

Однако стоит заметить, что люди с доходом 50 001 - 200 000 берут кредиты чаще, чем люди с более высоким доходом.
    
</div>

<p>
<div class="alert-danger">
<b>Комментарий ревьюера ❌</b> 
    
Нужно сделать сводную таблицу, где строки – категория, значения – это наличие долга, а столбцы – число всех кредитополучателей, количество должников и доля должников. Именно по доле должников мы и делаем дальнейшие выводы. При этом нужно так же оценить размеры групп, нет ли перекоса.

<p>
<div class="alert-info">
<b>Комментарий студента</b>

**Вывод:** 

Чем выше доход кредитополучателя, тем чаще он будет выплачивать его в срок. Однако, данные несбалансированы, поэтому исследовались две самые большие категории. Было замечено, что люди с доходом 50 001 - 200 000 берут кредиты чаще, чем люди с более высоким доходом 200 001 - 1 000 000.
    
</div>

<p>
<div class="alert-danger">
<b>Комментарий ревьюера ❌</b> 

    
Здесь нужно обратить внимание, что выборки крайне несбалансированны и для выявления зависимости стоит сравнить две самые большие категории B и C.

<p>
<div class="alert-success">
<b>Комментарий ревьюера 2 ✔️</b>

    
Верно 👍</div>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

В датафрейме был создан ноывй столбец `purpose_category`, в котором для удобства цели кредита были разбиты на категории по ключевым словам.

В столбце `purpose_category` указаны следующие категории в зависимости от цели кредита:

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

<p>
<div class="alert-info">
<b>Комментарий студента</b>
    
Создадим сводную таблицу, где:
* строки - категория в зависимости от цели кредита; 
* колонка 'count debt' - число кредитополучателей;
* колонка 'sum debt' - количество должников;
* колонка 'mean debt' - доля должников.
    
</div>

In [26]:
data.pivot_table(index='purpose_category', values='debt', aggfunc=['count', 'sum', 'mean'])

count  sum      mean
                           debt debt      debt
purpose_category                              
операции с автомобилем     4279  400  0.093480
операции с недвижимостью  10751  780  0.072551
получение образования      3988  369  0.092528
проведение свадьбы         2313  183  0.079118

<p>
<div class="alert-info">
<b>Комментарий студента</b>
    
Наибольшая доля должников - люди, берущие кредит на операции с автомобилем	и для  получение образования (9.3%).

Меньше всего должников среди людей, берущих кредит на покупку недвижимости (7.3%) или проведение свадьбы (7.9%).

Однако стоит отметить перекос в группах. Несмотря на то, что меньше всего должников среди людей, берущих кредит на проведение свадьбы, эти люди реже всех берут кредиты. Больше всего кредит берут для операций с недвижимостью, кредитополучателей с такой целью почти 10 800.
    
</div> 


<p>
<div class="alert-danger">
<b>Комментарий ревьюера ❌</b> 
    
Нужно сделать сводную таблицу, где строки – категория, значения – это наличие долга, а столбцы – число всех кредитополучателей, количество должников и доля должников. Именно по доле должников мы и делаем дальнейшие выводы. При этом нужно так же оценить размеры групп, нет ли перекоса.

<p>
<div class="alert-info">
<b>Комментарий студента</b>

**Вывод:** 

Люди, которые берут кредит для покупки недвижимости, чаще выплачивают его в срок. Также редкие должники люди, берущие кредит на проведение свадьбы, однако люди с такой целью не часто берут кредит. Чаще всех кредиты берут люди с целью приобрести недвижимость.
    
</div> 

<p>
<div class="alert-success">
<b>Комментарий ревьюера 2 ✔️</b>

    
Молодец 👍</div>

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 

В исходных данных пропуски были в столбцах с информацией об общем трудовом стаже и ежемесячном доходе. Данные представлены банком. 

Предполагаю, что данные взяты из анкет или заявок, которые оформляли люди, желающие получить кредит. Пропуски в колонке с ежемесячным доходом могут быть связаты с тем, что люди не захотели раскрывать свой доход. Также возможно это человеческий фактор, необходимые данные случайно не внесли. 

Пропуски в столбце общего трудового стажа могут быть связаны как с человеческим фактором, так и сложностями расчета этого показателя.

<div class="alert-warning">
<b>Комментарий ревьюера ⚠️</b>

    
Помимо человеческого фактора (кто-то что-то не ввел, умышленно или нет) не стоит исключать и технические ошибки: неправильный запрос к БД, некорректная выгрузка и т.п.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 

Медиана дает более реалистичную оценку, особенно если в значениях есть выбросы. А среднее значение некорректно характеризует данные, и при заполнении пропусков средним, будет характерное смещение. 

<p>
<div class="alert-success">
<b>Комментарий ревьюера ✔️</b>

Действительно, в данных могут наблюдаться некоторые "выбросы", и даже небольшое их количество может сильно "перетягивать" среднее значение в большую или меньшую сторону. Медиана зачастую даёт более правдивую картину. Также важно в каждом случае смотреть на то, как распределены данные. 
</div>

### Шаг 4: общий вывод.

<p>
<div class="alert-info">
<b>Комментарий студента</b>

**Вывод:** 

В данном проекте исследовалась надежность заемщиков. А именно какие параметры влияют на факт погашения кредита в срок. Заказчик - кредитный отдел банка.

Данные для анализа состоят из таблицы с 12 столбцами. 

**Описание данных:**
* children — количество детей в семье
* days_employed — общий трудовой стаж в днях
* dob_years — возраст клиента в годах
* education — уровень образования клиента
* education_id — идентификатор уровня образования
* family_status — семейное положение
* family_status_id — идентификатор семейного положения
* gender — пол клиента
* income_type — тип занятости
* debt — имел ли задолженность по возврату кредитов
* total_income — ежемесячный доход
* purpose — цель получения кредита
    
В ходе работы была проведена предобработка данных и исследовательский анализ данных.
    
**Во время предобработки данных:**
1. В столбце 'days_employed' были найдены артефакты (аномалии) — отрицательное количество дней трудового стажа. Все отрицательные значения были заменены положительными значениями.
    
2. В столбцах 'days_employed' и 'total_income' были обнаружены пропуски 2174 значений. Эти пропуски были заполнены. Для столбца 'days_employed' пропуски заполнили медианным значением трудового стажа в зависимости от типа занятости человека. Для столбца 'total_income' пропуски заполнили медианным значением ежемесячного дохода в зависимости от типа занятости человека.
    
3. В столбце 'children' было найдено два аномальных значения. Были удалены строки, в которых встречаются такие аномальные значения.
4. В столбце 'total_income' был изменен тип данных с вещественного на целочисленный.
5. В столбце 'education' были удалены неявные дубликаты.
6. Был создан  столбец 'total_income_category' на основании диапазонов дохода. Были выделены следующие категории дохода:
- 'E' доход 0 - 30 000;
- 'D' доход 30 001 - 50 000;
- 'C' доход 50 001 - 200 000;
- 'B' доход 200 001 - 1 000 000;
- 'A' доход выше 1 000 000.
7. Был создан  столбец 'purpose_category' на основании целей кредита, указанных в столбце 'purpose'. Были выделены следующие категории:
- 'операции с автомобилем';
- 'операции с недвижимостью';
- 'проведение свадьбы';
- 'получение образования'.
    
**Во время исследовательского анализы были сформулированы следующие выводы:**
1. Чем меньше детей в семье, тем выше кредитоспособность семьи;
2. Меньше всего должников среди людей в разводе и состоящих в браке. Также редкие должники вдовцы/вдовы, однако они и не часто берут кредиты. Чаще всех кредиты берут люди в браке;
3. Чем выше доход кредитополучателя, тем чаще он будет выплачивать его в срок. Однако стоит заметить, что люди с доходом 50 001 - 200 000 берут кредиты чаще, чем люди с более высоким доходом 200 001 - 1 000 000;
4. Люди, которые берут кредит для покупки недвижимости, чаще выплачивают его в срок. Также редкие должники люди, берущие кредит на проведение свадьбы, однако люди с такой целью не часто берут кредит. Чаще всех кредиты берут люди с целью приобрести недвижимость.
    
**Рекомендации:**
- внимательнее заполнять данные о трудовом стаже и ежемесячном доходе. Пропуски могли нести в себе важную информацию для анализа;
- при составлении модели кредитного скоринга комплексно уделить внимание следующим показателям: количество детей в семье, уровень дохода семьи, семейное положение. Учитывать, что есть определенные группы людей, которые чаще берут кредит. Таким семьям уделить больше внимания при анализе.

<p>
<div class="alert-danger">
<b>Комментарий ревьюера ❌</b> 
    
После внесения исправлений, нужно поправить вывод. Вывод - это полный отчет о проделанной работе. Он должен содержать все: от обработки данных до рекомендаций заказчику. Я бы советовала в начале кратко описывать, с какими данными мы работали и что сделали, а затем отвечать на вопрос исследования - в нашем случае, рекомендовать заказчику, как формировать модель кредитного скоринга

<p>
<div class="alert-success">
<b>Комментарий ревьюера 2 ✔️</b>

    
Молодец, хороший подробный вывод 👍</div>

<div style="border:solid green 3px; padding: 20px">

**Общий вывод по проекту**

Екатерина, спасибо за проделанную работу. Мне понравился твой подход и в целом код работает верно. Есть несколько важных моментов, которые я отметила и которые можно было бы улучшить, подсказки ты найдешь в моих комментариях. 


**Что нужно переделать, чтобы я приняла твой проект🤓**:
 
* Сделать сводные таблицы в каждом пункте с расчётами.
* Обрати внимание, что выборки по уровню дохода крайне несбалансированны и для выявления зависимости стоит сравнить две самые большие категории B и C.
* После внесения изменений поправить выводы
    

**Рекомендации ты найдешь в моих комментариях!**
    
Полезная ссылка про графики:

https://habr.com/ru/post/468295/

и еще одна

https://pyprog.pro/mpl/mpl_short_guide.html


**Желаю удачи и жду твой проект на повторное ревью!**😉

<div style="border:solid green 3px; padding: 20px">

**Общий вывод по проекту 2**

Екатерина, спасибо за комментарии и исправления! Отличная работа! Поздравляю со сдачей проекта!  🎉 🎉 🎉

Советую тебе бесплатный курс по статистике, можно посмотреть отдельно интересующие тебя темы:
    
https://stepik.org/course/76/